File: script/java/app/src/main/java/org/example/puzzle/elementary/Q11.java
package org.example.puzzle.elementary;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Q11 {
  public static void main(String[] args) {
    final var begin = System.currentTimeMillis();

    final var n = new ArrayList<List<BigInteger>>();

    for (BigInteger a = BigInteger.ONE, b = BigInteger.ONE, c; n.size() < 11; a = b, b = c) {
      c = a.add(b);

      BigInteger x = Arrays.asList(c.toString().split("")).stream()
          .map(BigInteger::new)
          .reduce(BigInteger.ZERO, BigInteger::add);

      if (c.remainder(x).equals(BigInteger.ZERO)) {
        n.add(List.of(c, x, c.divide(x)));
      }
    }

    n.forEach(System.out::println);

    final var finish = System.currentTimeMillis();
    System.out.println((finish - begin) + "ms");
  }
}

File: script/node/src/puzzle/elementary/q11.ts
/**
 * フィボナッチ数列のうち、各桁の数値を足した数で割り切れる数を
 * 以下の例に続けて小さい方から５つ求めよ
 */

{
  // # number の場合
  const begin = performance.now();

  const n: number[][] = [];

  for (let a = 1, b = 1, c: number; n.length < 11; a = b, b = c) {
    c = a + b;

    const x = c
      .toString()
      .split('')
      .map(Number)
      .reduce((x, y) => x + y);

    if (c % x === 0) {
      n.push([c, x, c / x]);
    }
  }

  console.table(n);

  const finish = performance.now();
  console.log(`${finish - begin}ms`);

  // ★ 11個目
  // 679,891,637,638,612,200 / 84 = 8,093,948,067,126,336
  //   => Number.MAX_SAFE_INTEGER を超えるので正しい数値になっていない
}

{
  // # bigint の場合
  const begin = performance.now();

  const n: bigint[][] = [];

  for (let a = 1n, b = 1n, c: bigint; n.length < 11; a = b, b = c) {
    c = a + b;

    const x = c
      .toString()
      .split('')
      .map(BigInt)
      .reduce((x, y) => x + y);

    if (c % x === 0n) {
      n.push([c, x, c / x]);
    }
  }

  console.table(n);

  const finish = performance.now();
  console.log(`${finish - begin}ms`);

  // ★ 11個目
  // 160,500,643,816,367,088 / 72 = 2,229,175,608,560,654
}
